syntax = "proto3";

package fetch.oef.pb;

import "google/protobuf/struct.proto";

message StrIntPair {
    string first = 1;
    int32 second = 2;
}

message StrDoublePair {
    string first = 1;
    double second = 2;
}

message StrStrPair {
    string first = 1;
    string second = 2;
}

message TACController {

    message Registered {
    }
    message Unregistered {
    }
    message Cancelled {
    }

    message GameData {
        repeated StrIntPair amount_by_currency = 1;
        repeated StrDoublePair exchange_params_by_currency = 2;
        repeated StrIntPair quantities_by_good_pbk = 3;
        repeated StrDoublePair utility_params_by_good_pbk = 4;
        int64 tx_fee = 5;
        repeated StrStrPair agent_pbk_to_name = 6;
        repeated StrStrPair good_pbk_to_name = 7;
        string version_id = 8;
    }

    message TransactionConfirmation {
        string transaction_id = 1;
        repeated StrIntPair amount_by_currency = 2;
        repeated StrIntPair quantities_by_good_pbk = 3;
    }

    message StateUpdate {
        GameData game_data = 1;
        repeated TACAgent.Transaction txs = 2;
    }

    message Error {
        enum ErrorCode {
            GENERIC_ERROR = 0;
            REQUEST_NOT_VALID = 1;
            AGENT_PBK_ALREADY_REGISTERED = 2;
            AGENT_NAME_ALREADY_REGISTERED = 3;
            AGENT_NOT_REGISTERED = 4;
            TRANSACTION_NOT_VALID = 5;
            TRANSACTION_NOT_MATCHING = 6;
            AGENT_NAME_NOT_IN_WHITELIST = 7;
            COMPETITION_NOT_RUNNING = 8;
            DIALOGUE_INCONSISTENT = 9;
        }

        ErrorCode error_code = 1;
        string error_msg = 2;
        google.protobuf.Struct details = 3;
    }

}

message TACAgent {

    message Register {
        string agent_name = 1;
    }
    message Unregister {
    }

    message Transaction {
        string transaction_id = 1;
        string counterparty = 2;
        repeated StrIntPair amount_by_currency = 3;
        int64 sender_tx_fee = 4;
        int64 counterparty_tx_fee = 5;
        repeated StrIntPair quantities_by_good_pbk = 6;
    }

    message GetStateUpdate {
    }

}

message TACMessage {
    oneof content{
        TACAgent.Register register = 1;
        TACAgent.Unregister unregister = 2;
        TACAgent.Transaction transaction = 3;
        TACAgent.GetStateUpdate get_state_update = 4;
        TACController.Registered registered = 5;
        TACController.Unregistered unregistered = 6;
        TACController.Cancelled cancelled = 7;
        TACController.GameData game_data = 8;
        TACController.TransactionConfirmation transaction_confirmation = 9;
        TACController.StateUpdate state_update = 10;
        TACController.Error error = 11;
    }
}
